pkg/admission/provider.go

```diff
@@ -1,93 +1,91 @@
 // TODO: only supports plain text with no auth, add security later
-// TODO: use keystone to access Neutron API
+// TODO: check keystone expiration date 
 package admission
 
 import (
-	resty "gopkg.in/resty.v1"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+)
+
+const (
+	providerDefaultRegion  = "RegionOne"
+	providerNeutronService = "neutron"
+	providerVersion        = "v2.0"
 )
 
 type Network struct {
-	Name    string
-	ID      string
-	Physnet string
+	Name    string `json:"name"`
+	ID      string `json:"id"`
+	Physnet string `json:"provider:physical_network"`
 }
 
 type providerClient struct {
-	client *resty.Client
+	client *gophercloud.ServiceClient
 }
 
 // NewProviderClient creates a REST client to access Neutron API
-func NewProviderClient(url string) *providerClient {
-	client := resty.
-		New().
-		SetHostURL(url).
-		SetHeader("Accept", "application/json")
-	return &providerClient{client}
+func NewProviderClient(keystoneUrl string, username string, password string) (*providerClient, error) {
+	tokensUrl := fmt.Sprintf("%s%s", keystoneUrl, providerVersion)
+
+	authOpts := gophercloud.AuthOptions{IdentityEndpoint: tokensUrl, Username: username, Password: password, TenantName: "admin"}
+	provider, err := openstack.AuthenticatedClient(authOpts)
+
+	if err != nil {
+		return nil, err
+	}
+
+	endpointOpts := gophercloud.EndpointOpts{Region: providerDefaultRegion, Name: providerNeutronService}
+	client, err := openstack.NewNetworkV2(provider, endpointOpts)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &providerClient{client}, nil
 }
 
 // ListNetworkIDsByNames returns a map where key is name of Neutron Network and key is its ID
 func (c *providerClient) ListNetworkByName() (map[string]*Network, error) {
-	var result map[string][]map[string]interface{}
 
-	// TODO: check provider error output
-	_, err := c.client.R().
-		SetResult(&result).
-		Get("v2.0/networks")
+	var networkList []Network
+
+	pages, err := networks.List(c.client, networks.ListOpts{}).AllPages()
+	if err != nil {
+		return nil, err
+	}
+
+	err = networks.ExtractNetworksInto(pages, &networkList)
```

Comment:
I am unsure if I understand the magic happens here.
Is the idea to have your own Network data type, which is independent of the OpenStack library?
And this line converts the json into your own data type?
And this way only this single file depends on the OpenStack library, but not the file which consumes the Networks?

Reply:
Yes. I have to use own Network struct because the library one is missing the "provider:physical_network". I'll report issue in the library and create PR for this so we can fully use the library without the need for our own struct.